pallets/assets-handler/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+//! A pallet for handling bridge transfer lgoic.
+#![cfg_attr(not(feature = "std"), no_std)]
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		tokens::{
+			fungible::{Inspect as FInspect, Mutate as FMutate},
+			fungibles::Mutate as FsMutate,
+			Fortitude, Precision,
+		},
+		StorageVersion,
+	},
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use parity_scale_codec::FullCodec;
+type BalanceOf<T> =
+	<pallet_balances::Pallet<T> as FInspect<<T as frame_system::Config>::AccountId>>::Balance;
+use pallet_bridge_transfer::BridgeHandler;
+use pallet_parachain_staking::IssuanceAdapter;
+use sp_runtime::{
+	traits::{CheckedAdd, CheckedSub},
+	ArithmeticError, DispatchError, FixedPointOperand,
+};
+use sp_std::{fmt::Debug, prelude::*};
+type ResourceId = pallet_bridge::ResourceId;
+
+#[derive(PartialEq, Eq, Clone, Encode, Debug, Decode, TypeInfo)]
+pub struct AssetInfo<AssetId, Balance> {
+	pub fee: Balance,
+	// None for native token
+	pub asset: Option<AssetId>,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+	type AssetId<T> = <T as pallet_assets::Config>::AssetId;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ pallet_balances::Config
+		+ pallet_assets::Config
+		+ pallet_bridge::Config
+	{
+		/// Overarching event type
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Treasury account to receive assets fee
+		type TreasuryAccount: Get<Self::AccountId>;
+
+		/// The privileged origin to call update_maximum_issuance
+		type SetMaximumIssuanceOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		#[pallet::constant]
+		type DefaultMaximumIssuance: Get<BalanceOf<Self>>;
+
+		#[pallet::constant]
+		// In parachain local decimal format
+		type ExternalTotalIssuance: Get<BalanceOf<Self>>;
+	}
+
+	// Resource Id of pallet assets token
+	#[pallet::storage]
+	#[pallet::getter(fn resource_to_asset_info)]
+	pub type ResourceToAssetInfo<T: Config> =
+		StorageMap<_, Twox64Concat, ResourceId, AssetInfo<AssetId<T>, BalanceOf<T>>, OptionQuery>;
+
+	#[pallet::type_value]
+	pub fn DefaultExternalBalances<T: Config>() -> BalanceOf<T> {
+		T::ExternalTotalIssuance::get()
+			.checked_sub(&pallet_balances::Pallet::<T>::total_issuance())
+			.map_or_else(|| 0u32.into(), |v| v)
+	}
+
+	// Native Token External Unlocked Total Balances of outside
+	#[pallet::storage]
+	#[pallet::getter(fn external_balances)]
+	pub type ExternalBalances<T: Config> =
+		StorageValue<_, BalanceOf<T>, ValueQuery, DefaultExternalBalances<T>>;
```

Comment:
I sort of dont' understand/forget why we need `ExternalBalances` at all - it seems to be always equal to total LIT circulation - total issuance on this chain, unless you explicitly set it.

But we won't set it because it will affect the inflation in pallet-parachain-staking.

If we only want to limit the maximal tokens that can be minted, we have `MaximumIssuance` already.

Reply:
External Balances = total supply of non-local native token. And yes, it is supposed to target at around a total sum of 100 Million LIT. Which is used by parachain staking.

MaximumIssuance is for how much local bridge can be minted. No matter how much outside LIT want to get in parachain, at most this amount of LIT will be allowed to minted.

They are different.